### PR TITLE
graalvmce: 21.0.0 -> 21.2.0

### PR DESCRIPTION
**Pull request by ericdallo**
###### Motivation for this change
Bump graalvmce

###### Things done
- Removed support for graalvmce-java8 for darwin as is not supported anymore

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ericdallo** [comment]
@GrahamcOfBorg build babashka clj-kondo clojure-lsp graalvm8-ce graalvm11-ce

**Comment by ericdallo** [comment]
Result of `nixpkgs-review pr 120072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>babashka</li>
    <li>clojure-lsp</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>clj-kondo</li>
    <li>graalvm11-ce</li>
    <li>graalvm8-ce</li>
  </ul>
</details>

**Comment by ericdallo** [comment]
@GrahamcOfBorg eval

**Comment by ericdallo** [comment]
@SuperSandro2000 not sure if it's common for eval fail since there is no graalvm8-ce for darwin.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -10850,9 +10850,12 @@ in
     inherit (darwin) libiconv libobjc libresolv;
   }) mx jvmci8 graalvm8;
 
-  inherit (callPackages ../development/compilers/graalvm/community-edition.nix {
+  graalvm-ce-pkgs = callPackages ../development/compilers/graalvm/community-edition.nix {
     inherit (darwin.apple_sdk.frameworks) Foundation;
-  }) graalvm8-ce graalvm11-ce;
+  };
+
+  graalvm8-ce = if stdenv.isLinux then graalvm-ce-pkgs.graalvm8-ce else null;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:10857]: That is not allowed because packages can never be null. Please mark it broken instead.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
> not sure if it's common for eval fail since there is no graalvm8-ce for darwin.

That is because packages are never allowed to be null.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (ericdallo) [reply pkgs/top-level/all-packages.nix:10857]: even if linux works?
The change on this PR is that this new version of graalvm doesn't support darwin anymore for java8 but it supports for java11 and both for linux

Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:10857]: > even if linux works?

You can mark it broken on darwin or non linux, too.

**Comment by ericdallo** [comment]
@SuperSandro2000 now it's failling because it tries to download the wrong version, is there any way to throw an exception or force fail if the platform is X since I already included `broken = (stdenv.isDarwin && javaVersion == "8");`

**Comment by SuperSandro2000** [comment]
I think you would need to exclude that platform conditionally. Then the broken can be removed, too.

**Comment by SuperSandro2000** [comment]
Still not working. Not sure how to solve this. Maybe we need to split the versions.

**Comment by ericdallo** [comment]
Yeah, that's what I thought, I'll try that

**Comment by thiagokokada** [comment]
BTW, needs this PR merged to bump `babashka` to version `0.4.6`.

**Comment by ericdallo** [comment]
@thiagokokada I'm having issues on the current graalvmce derivation arch, we need to remove the macos support on graalvmce8, but this derivation uses a generic approach, we need to think in a better way to organize that

**Comment by ericdallo** [comment]
@GrahamcOfBorg eval

**Comment by ericdallo** [comment]
Fixed the issue @SuperSandro2000 Can you re-check it please?

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -11633,9 +11633,12 @@ in
...
 
-  inherit (callPackages ../development/compilers/graalvm/community-edition.nix {
-    inherit (darwin.apple_sdk.frameworks) Foundation;
-  }) graalvm8-ce graalvm11-ce;
+  graalCEPackages =
+    recurseIntoAttrs (callPackage ../development/compilers/graalvm/community-edition {
+      inherit (darwin.apple_sdk.frameworks) Foundation;
+    });
+  graalvm8-ce = graalCEPackages.graalvm8-ce;
+  graalvm11-ce = graalCEPackages.graalvm11-ce;
```
Comment (babariviere) [inline pkgs/top-level/all-packages.nix:11641]:
```suggestion
  graalvmCEPackages =
    recurseIntoAttrs (callPackage ../development/compilers/graalvm/community-edition {
      inherit (darwin.apple_sdk.frameworks) Foundation;
    });
  graalvm8-ce = graalvmCEPackages.graalvm8-ce;
  graalvm11-ce = graalvmCEPackages.graalvm11-ce;
```

**Comment by babariviere** [comment]
Result of `nixpkgs-review pr 120072` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>babashka</li>
    <li>clojure-lsp</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>clj-kondo</li>
    <li>graalvm11-ce (graalCEPackages.graalvm11-ce)</li>
    <li>zprint</li>
  </ul>
</details>

**Review by babariviere — APPROVED** [review]

**Review by babariviere — APPROVED** [review]

**Comment by ericdallo** [comment]
@babariviere made one more change, fixing babashka reflection, now all binaries should compile successfully :)

**Comment by babariviere** [comment]
@ericdallo Ok, I will recompile everything on darwin to see if it works 👍🏻

File: pkgs/development/interpreters/clojure/babashka.nix [hunk 3]
```diff
@@ -52,6 +52,7 @@ stdenv.mkDerivation rec {
           "--no-server"
           "--report-unsupported-elements-at-runtime"
           "--initialize-at-run-time=org.postgresql.sspi.SSPIClient"
+          "--initialize-at-run-time=org.httpkit.client.ClientSslEngineFactory\$SSLHolder"
```
Comment (thiagokokada) [inline pkgs/development/interpreters/clojure/babashka.nix:55]: Maybe it would be better to separate this on another commit.

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{ callPackage, Foundation }:
+
+let
+  mkGraal = opts: callPackage (import ./repository.nix opts) {
+    Foundation = Foundation;
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:5]:
```suggestion
    inherit Foundation;
```

**Comment by babariviere** [comment]
Result of `nixpkgs-review pr 120072` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>graalvm11-ce</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  };
+in {
+  mkGraal = mkGraal;
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:8]:
```suggestion
  inherit mkGraal;
```

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 120072` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>graalvm11-ce (graalvmCEPackages.graalvm11-ce)</li>
    <li>graalvm8-ce (graalvmCEPackages.graalvm8-ce)</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Comment by ericdallo** [comment]
Done @thiagokokada

**Review by babariviere — APPROVED** [review]

File: pkgs/development/compilers/graalvm/community-edition/default.nix [hunk 6]
```diff
@@ -0,0 +1,21 @@
+{ callPackage, Foundation }:
+
+let
+  mkGraal = opts: callPackage (import ./repository.nix opts) {
+    inherit Foundation;
+  };
+in {
+  inherit mkGraal;
```
Comment (thiagokokada) [inline pkgs/development/compilers/graalvm/community-edition/default.nix:8]: It doesn't seem this is being used. Can't it be deleted :thinking: ?

**Review by thiagokokada — APPROVED** [review]

File: pkgs/development/compilers/graalvm/community-edition/default.nix (hunk 6 above)
Comment (ericdallo) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:8]: this make it possible to any derivation use it's own graalvm version, like we do for mkFlutter, with that we can have babashka using a graalvm different from clojure-lsp for example

Comment (thiagokokada) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:8]: Yeah, but this isn't exposed on `all-packages.nix`. So unless I misunderstood, this can't be used for this purpose.

Comment (ericdallo) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:8]: I'll expose it on all-packages.nix so

Comment (ericdallo) [reply pkgs/development/compilers/graalvm/community-edition/default.nix:8]: @thiagokokada we expose the `graalvmCEPackages`, so it would be possible use like `graalvmCEPackages.mkGraal` right? just like the flutter one

**Comment by ericdallo** [comment]
@SuperSandro2000 Are we good to merge this?
